lib/fxpay/adapter.js

```diff
@@ -37,28 +37,39 @@
     // This must execute callback(error) when finished.
     // The error parameter should be null when there are no errors.
     //
-    if (!settings.hasAddReceipt && !settings.localStorage) {
-      settings.log.error('no way to store receipts on this platform');
-      return callback('PAY_PLATFORM_UNAVAILABLE');
-    }
-    if (settings.appSelf &&
-        settings.window.location.href.indexOf('app://') === 0 &&
-        !settings.appSelf.manifest.origin) {
-      settings.log.error('packaged app did not define an origin so ' +
-                         'we have no key to look up products');
-      return callback('UNDEFINED_APP_ORIGIN');
-    }
-    var numReceipts = 0;
-    var receipt;
-    var allReceipts = receipts.all();
-    for (var i = 0; i < allReceipts.length; i++) {
-      receipt = allReceipts[i];
-      settings.log.info('Installed receipt: ' + receipt);
-      numReceipts++;
-      receipts.validateInAppProductReceipt(receipt, settings.onrestore);
-    }
-    settings.log.info('Number of receipts installed: ' + numReceipts);
-    callback();
+    utils.getAppSelf(function(error, appSelf) {
+      if (error) {
+        return callback(error);
+      }
+
+      var platformHasAddReceipt = appSelf && appSelf.addReceipt;
+      if (!platformHasAddReceipt && !settings.localStorage) {
+        settings.log.error('no way to store receipts on this platform');
+        return callback('PAY_PLATFORM_UNAVAILABLE');
+      }
+
+      if (appSelf && settings.window.location.href.indexOf('app://') === 0 &&
+          !appSelf.manifest.origin) {
+        settings.log.error('packaged app did not define an origin so ' +
+                           'we have no key to look up products');
+        return callback('UNDEFINED_APP_ORIGIN');
+      }
+
+      receipts.all(function(error, allReceipts) {
+        if (error) {
+          return callback(error);
+        }
+        var numReceipts = 0;
+
+        allReceipts.forEach(function(receipt) {
+          settings.log.info('Installed receipt: ' + receipt);
+          numReceipts++;
```

Comment:
I see that this didn't change but it looks like it could use `allReceipts.length` when logging instead of counting.